File: baekjoon/[1029] 무작위_Gold3-5/[1062] 가르침.cpp
#include <bits/stdc++.h>
using namespace std;
int N, K, result = 0;
int arr[50] = { 0 };
void dfs(int cnt, int status, int j) {
    if (cnt == K-5) {
        int temp = 0;
        for (int i = 0; i < N; i++) {
            if ((status | arr[i]) == status)
                temp++;
        }
        result = max(result, temp);
        return;
    }
    for (int i = j; i <= 25; i++) {
        if (status & (1 << i)) continue;
        status |= 1 << i;
        dfs(cnt + 1, status, i+1);
        status &= ~(1 << i);
    }
}
int main() {
    cin >> N >> K;
    string s;
    for (int i = 0; i < N; i++) {
        cin >> s;
        for (int j = 0; j < s.size(); j++) {
            arr[i] |= 1 << (s[j] - 'a');
        }
    }

    if (K < 5)
        cout << 0;
    else {
        int temp = 0;
        temp |= 1 << ('a' - 'a');
        temp |= 1 << ('n' - 'a');
        temp |= 1 << ('t' - 'a');
        temp |= 1 << ('i' - 'a');
        temp |= 1 << ('c' - 'a');
        dfs(0, temp, 0);
        cout << result;
    }
    return 0;
}

File: baekjoon-java/[1062] 가르침.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

import static java.lang.Math.max;

public class Main {
    static int N, K, W, ANS = 0;
    static int[] arr;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());

        arr = new int[N];
        for (int i = 0; i<N; i++){
            st = new StringTokenizer(br.readLine());
            String str = st.nextToken();
            for (int j = 0; j<str.length(); j++){
                arr[i]|=(1<<(str.charAt(j)-'a'));
            }
        }

        solve(0,0);
        System.out.println(ANS);
    }

    static void solve(int cnt, int start){
        if (cnt==K){
            int result = 0;
            for (int i =0; i<N; i++){
                if ((arr[i]|W)==W) result++;
            }
            ANS = max(ANS, result);
            return;
        }
        for (int i = start; i<26; i++){
            if ((W&(1<<i))!=0) continue;
            W|=(1<<i);
            solve(cnt+1, i+1);
            W&=~(1<<i);
        }
    }
}
